BUG: UJSON4C fails to decode answer

This bug appears sporadically; so far not consistently reproducible.

UJSON4C library's `UJDecode()` fails to decode received JSON answer; the answer is correct. The failure is always `Trailing data`. 
Sometimes the answer buffer gets corrupted after the function returns, although no writing to it seems to occur in the library.

Log excerpts:

1.
```
Mon Nov 19 09:57:26 2018 - [DEBUG] post_json()@connect.c:568 [STMT@0x000001CBF9FA6860] POSTing JSON [38] `{"query": "SYS TYPES", "mode": "ODBC"}`.
[...]
Mon Nov 19 09:57:26 2018 - [DEBUG] attach_answer()@queries.c:223 [STMT@0x000001CBF9FA6860] attaching answer [2324]`{"columns":[{"name":"TYPE_NAME","type":"keyword"},{"name":"DATA_TYPE","type":"integer"},{"name":"PRECISION","type":"integer"},{"name":"LITERAL_PREFIX","type":"keyword"},{"name":"LITERAL_SUFFIX","type":"keyword"},{"name":"CREATE_PARAMS","type":"keyword"},{"name":"NULLABLE","type":"short"},{"name":"CASE_SENSITIVE","type":"boolean"},{"name":"SEARCHABLE","type":"short"},{"name":"UNSIGNED_ATTRIBUTE","type":"boolean"},{"name":"FIXED_PREC_SCALE","type":"boolean"},{"name":"AUTO_INCREMENT","type":"boolean"},{"name":"LOCAL_TYPE_NAME","type":"keyword"},{"name":"MINIMUM_SCALE","type":"short"},{"name":"MAXIMUM_SCALE","type":"short"},{"name":"SQL_DATA_TYPE","type":"integer"},{"name":"SQL_DATETIME_SUB","type":"integer"},{"name":"NUM_PREC_RADIX","type":"integer"},{"name":"INTERVAL_PRECISION","type":"integer"}],"rows":[["BYTE",-6,3,"'","'",null,2,false,3,false,false,false,null,0,0,-6,0,10,null],["LONG",-5,19,"'","'",null,2,false,3,false,false,false,null,0,0,-5,0,10,null],["BINARY",-3,2147483647,"'","'",null,2,false,3,true,false,false,null,null,null,-3,0,null,null],["NULL",0,0,"'","'",null,2,false,3,true,false,false,null,null,null,0,0,null,null],["INTEGER",4,10,"'","'",null,2,false,3,false,false,false,null,0,0,4,0,10,null],["SHORT",5,5,"'","'",null,2,false,3,false,false,false,null,0,0,5,0,10,null],["HALF_FLOAT",6,16,"'","'",null,2,false,3,false,false,false,null,0,16,6,0,2,null],["SCALED_FLOAT",6,19,"'","'",null,2,false,3,false,false,false,null,0,19,6,0,2,null],["FLOAT",7,7,"'","'",null,2,false,3,false,false,false,null,0,7,7,0,2,null],["DOUBLE",8,15,"'","'",null,2,false,3,false,false,false,null,0,15,8,0,2,null],["KEYWORD",12,256,"'","'",null,2,true,3,true,false,false,null,null,null,12,0,null,null],["TEXT",12,2147483647,"'","'",null,2,true,3,true,false,false,null,null,null,12,0,null,null],["IP",12,39,"'","'",null,2,false,3,true,false,false,null,null,null,12,0,null,null],["BOOLEAN",16,1,"'","'",null,2,false,3,true,false,false,null,null,null,16,0,null,null],["DATE",93,24,"'","'",null,2,false,3,true,false,false,null,3,3,9,3,null,null],["UNSUPPORTED",1111,0,"'","'",null,2,false,3,true,false,false,null,null,null,1111,0,null,null],["OBJECT",2002,0,"'","'",null,2,false,3,true,false,false,null,null,null,2002,0,null,null],["NESTED",2002,0,"'","'",null,2,false,3,true,false,false,null,null,null,2002,0,null,null]]}`.
Mon Nov 19 09:57:26 2018 - [ERROR] attach_answer()@queries.c:230 [STMT@0x000001CBF9FA6860] failed to decode JSON answer: Trailing data ([2324] `¤² p«U`).
```
2.
```
Mon Oct 22 10:00:30 2018 - [DEBUG] post_json()@connect.c:566 [STMT@0x0000018AA177EBC0] POSTing JSON [108] `{"query": "SYS COLUMNS CATALOG '7879474408f44e9ca98d1597e659dfe2' TABLE LIKE 'users' ESCAPE '\\'  LIKE '%'"}`.
[...]
Mon Oct 22 10:00:30 2018 - [DEBUG] attach_answer()@queries.c:222 [STMT@0x0000018AA177EBC0] attaching answer [1665]`{"columns":[{"name":"TABLE_CAT","type":"keyword"},{"name":"TABLE_SCHEM","type":"keyword"},{"name":"TABLE_NAME","type":"keyword"},{"name":"COLUMN_NAME","type":"keyword"},{"name":"DATA_TYPE","type":"integer"},{"name":"TYPE_NAME","type":"keyword"},{"name":"COLUMN_SIZE","type":"integer"},{"name":"BUFFER_LENGTH","type":"integer"},{"name":"DECIMAL_DIGITS","type":"integer"},{"name":"NUM_PREC_RADIX","type":"integer"},{"name":"NULLABLE","type":"integer"},{"name":"REMARKS","type":"keyword"},{"name":"COLUMN_DEF","type":"keyword"},{"name":"SQL_DATA_TYPE","type":"integer"},{"name":"SQL_DATETIME_SUB","type":"integer"},{"name":"CHAR_OCTET_LENGTH","type":"integer"},{"name":"ORDINAL_POSITION","type":"integer"},{"name":"IS_NULLABLE","type":"keyword"},{"name":"SCOPE_CATALOG","type":"keyword"},{"name":"SCOPE_SCHEMA","type":"keyword"},{"name":"SCOPE_TABLE","type":"keyword"},{"name":"SOURCE_DATA_TYPE","type":"short"},{"name":"IS_AUTOINCREMENT","type":"keyword"},{"name":"IS_GENERATEDCOLUMN","type":"keyword"}],"rows":[["7879474408f44e9ca98d1597e659dfe2",null,"users","password",12,"TEXT",0,2147483647,null,null,1,null,null,12,0,2147483647,1,"YES",null,null,null,null,"NO","NO"],["7879474408f44e9ca98d1597e659dfe2",null,"users","password.keyword",12,"KEYWORD",0,2147483647,null,null,1,null,null,12,0,2147483647,1,"YES",null,null,null,null,"NO","NO"],["7879474408f44e9ca98d1597e659dfe2",null,"users","user",12,"TEXT",0,2147483647,null,null,1,null,null,12,0,2147483647,2,"YES",null,null,null,null,"NO","NO"],["7879474408f44e9ca98d1597e659dfe2",null,"users","user.keyword",12,"KEYWORD",0,2147483647,null,null,1,null,null,12,0,2147483647,1,"YES",null,null,null,null,"NO","NO"]]}`.
Mon Oct 22 10:00:30 2018 - [ERROR] attach_answer()@queries.c:228 [STMT@0x0000018AA177EBC0] failed to decode JSON answer (`{"columns":[{"name":"TABLE_CAT","type":"keyword"},{"name":"TABLE_SCHEM","type":"keyword"},{"name":"TABLE_NAME","type":"keyword"},{"name":"COLUMN_NAME","type":"keyword"},{"name":"DATA_TYPE","type":"integer"},{"name":"TYPE_NAME","type":"keyword"},{"name":"COLUMN_SIZE","type":"integer"},{"name":"BUFFER_LENGTH","type":"integer"},{"name":"DECIMAL_DIGITS","type":"integer"},{"name":"NUM_PREC_RADIX","type":"integer"},{"name":"NULLABLE","type":"integer"},{"name":"REMARKS","type":"keyword"},{"name":"COLUMN_DEF","type":"keyword"},{"name":"SQL_DATA_TYPE","type":"integer"},{"name":"SQL_DATETIME_SUB","type":"integer"},{"name":"CHAR_OCTET_LENGTH","type":"integer"},{"name":"ORDINAL_POSITION","type":"integer"},{"name":"IS_NULLABLE","type":"keyword"},{"name":"SCOPE_CATALOG","type":"keyword"},{"name":"SCOPE_SCHEMA","type":"keyword"},{"name":"SCOPE_TABLE","type":"keyword"},{"name":"SOURCE_DATA_TYPE","type":"short"},{"name":"IS_AUTOINCREMENT","type":"keyword"},{"name":"IS_GENERATEDCOLUMN","type":"keyword"}],"rows":[["7879474408f44e9ca98d1597e659dfe2",null,"users","password",12,"TEXT",0,2147483647,null,null,1,null,null,12,0,2147483647,1,"YES",null,null,null,null,"NO","NO"],["7879474408f44e9ca98d1597e659dfe2",null,"users","password.keyword",12,"KEYWORD",0,2147483647,null,null,1,null,null,12,0,2147483647,1,"YES",null,null,null,null,"NO","NO"],["7879474408f44e9ca98d1597e659dfe2",null,"users","user",12,"TEXT",0,2147483647,null,null,1,null,null,12,0,2147483647,2,"YES",null,null,null,null,"NO","NO"],["7879474408f44e9ca98d1597e659dfe2",null,"users","user.keyword",12,"KEYWORD",0,2147483647,null,null,1,null,null,12,0,2147483647,1,"YES",null,null,null,null,"NO","NO"]]}`): Trailing data.
```
3.
```
Thu Oct 25 12:21:10 2018 - [DEBUG] post_json()@connect.c:566 [STMT@0x00000273EA3D3040] POSTing JSON [72] `{"query": "SYS TABLES CATALOG LIKE '%' ESCAPE '\\' TYPE 'TABLE','VIEW'"}`.
[...]
Thu Oct 25 12:21:10 2018 - [DEBUG] attach_answer()@queries.c:223 [STMT@0x00000273EA3D3040] attaching answer [775]`{"columns":[{"name":"TABLE_CAT","type":"keyword"},{"name":"TABLE_SCHEM","type":"keyword"},{"name":"TABLE_NAME","type":"keyword"},{"name":"TABLE_TYPE","type":"keyword"},{"name":"REMARKS","type":"keyword"},{"name":"TYPE_CAT","type":"keyword"},{"name":"TYPE_SCHEM","type":"keyword"},{"name":"TYPE_NAME","type":"keyword"},{"name":"SELF_REFERENCING_COL_NAME","type":"keyword"},{"name":"REF_GENERATION","type":"keyword"}],"rows":[["7879474408f44e9ca98d1597e659dfe2","",".kibana","TABLE","",null,null,null,null,null],["7879474408f44e9ca98d1597e659dfe2","",".security-6","TABLE","",null,null,null,null,null],["7879474408f44e9ca98d1597e659dfe2","","employ_name","TABLE","",null,null,null,null,null],["7879474408f44e9ca98d1597e659dfe2","","users","TABLE","",null,null,null,null,null]]}`.
Thu Oct 25 12:21:10 2018 - [ERROR] attach_answer()@queries.c:230 [STMT@0x00000273EA3D3040] failed to decode JSON answer: Trailing data ([775] `6ÈD$æõ`).
```

Spurious issue reoccurs:
```Tue Jan 29 16:27:21 2019 - [ERROR] attach_answer()@queries.c:241 [STMT@0x034059A0] failed to decode JSON answer: Trailing data ([214] `{"columns":[{"name":"1","type":"integer","display_size":11}],"rows":[[1],[1],[1],[1],[1]],"cursor":"o6CrAwFzQERYRjFaWEo1UVc1a1JtVjBZMmdCQUFBQUFBQUFBSlFXUjFSbWJIaEpWbFpTY1MweGNtazVhVTkzU210a1p3PT0FAQFwAWMBAAAAAQA="}`).```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -102,11 +102,6 @@ if (${WIN32})
 	# 2MB stack size
 	# set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /STACK:2097152")
 	# set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /F 2097152")
-	# Force UJSON4C's Unicode decoder scratchpad onto the heap. Its large
-	# default size (128K) leads to high stack usage.
-	set(CMAKE_C_FLAGS
-		# defined to sizeof(wchar_t)
-		"${CMAKE_C_FLAGS} /DJSON_MAX_STACK_BUFFER_SIZE=2")
 
 	if (${IS_UNICODE})
 		set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /DUNICODE /D_UNICODE")

--- a/driver/connect.c
+++ b/driver/connect.c
@@ -313,7 +313,13 @@ static size_t write_callback(char *ptr, size_t size, size_t nmemb,
 				}
 			}
 		}
-		wbuf = realloc(dbc->abuff, need);
+		/* Add a 0-term for UJSON4C: while the lib takes the string length to
+		 * parse, it's not really safe if the string is not a NTS. Adding the
+		 * 0-term at the end of the chunk/string will ensure the library won't
+		 * read past it (though this won't prevent it from failing to parse
+		 * valid JSON within the indicated length if there's white space in
+		 * the chunk right past the indicated JSON length.) */
+		wbuf = realloc(dbc->abuff, need + /*\0*/1);
 		if (! wbuf) {
 			ERRNH(dbc, "libcurl: failed to realloc to %zdB.", need);
 			return 0;
@@ -324,6 +330,8 @@ static size_t write_callback(char *ptr, size_t size, size_t nmemb,
 
 	memcpy(dbc->abuff + dbc->apos, ptr, have);
 	dbc->apos += have;
+	/* Add the 0-term for UJSON4C (but don't count it - see above) */
+	dbc->abuff[dbc->apos] = '\0';
 	DBGH(dbc, "libcurl: copied %zdB: `%.*s`.", have, have, ptr);
 
 	/*
@@ -1245,7 +1253,9 @@ static SQLRETURN check_server_version(esodbc_dbc_st *dbc)
 
 	obj = UJDecode(resp.str, resp.cnt, /*heap f()s*/NULL, &state);
 	if (! obj) {
-		ERRH(dbc, "failed to parse as JSON");
+		ERRH(dbc, "failed to parse JSON: %s ([%zd] `" LCPDL "`).",
+			state ? UJGetError(state) : "<none>",
+			resp.cnt, LCSTR(&resp));
 		goto err;
 	}
 	memset(&o_version, 0, sizeof(o_version));

